<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Identify biomarkers using supervised leaning (SL) methods — run_sl • MicrobiomeAnalysis</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Identify biomarkers using supervised leaning (SL) methods — run_sl"><meta property="og:description" content="Identify biomarkers using logistic regression, random forest, or support
vector machine."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MicrobiomeAnalysis</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Introduction.html">MicrobiomeAnalysis: A comprehensive R package for managing and analyzing microbiome data</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/HuaZou/MicrobiomeAnalysis/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Identify biomarkers using supervised leaning (SL) methods</h1>
    <small class="dont-index">Source: <a href="https://github.com/HuaZou/MicrobiomeAnalysis/blob/HEAD/R/da_supervised_leaning.R" class="external-link"><code>R/da_supervised_leaning.R</code></a></small>
    <div class="hidden name"><code>run_sl.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Identify biomarkers using logistic regression, random forest, or support
vector machine.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">run_sl</span><span class="op">(</span>
  <span class="va">ps</span>,
  <span class="va">group</span>,
  taxa_rank <span class="op">=</span> <span class="st">"all"</span>,
  transform <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"identity"</span>, <span class="st">"log10"</span>, <span class="st">"log10p"</span><span class="op">)</span>,
  norm <span class="op">=</span> <span class="st">"none"</span>,
  norm_para <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,
  nfolds <span class="op">=</span> <span class="fl">3</span>,
  nrepeats <span class="op">=</span> <span class="fl">3</span>,
  sampling <span class="op">=</span> <span class="cn">NULL</span>,
  tune_length <span class="op">=</span> <span class="fl">5</span>,
  top_n <span class="op">=</span> <span class="fl">10</span>,
  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"LR"</span>, <span class="st">"RF"</span>, <span class="st">"SVM"</span><span class="op">)</span>,
  <span class="va">...</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>ps</dt>
<dd><p>a <code><a href="https://rdrr.io/pkg/phyloseq/man/phyloseq-class.html" class="external-link">phyloseq-class</a></code> object.</p></dd>
<dt>group</dt>
<dd><p>character, the variable to set the group.</p></dd>
<dt>taxa_rank</dt>
<dd><p>character to specify taxonomic rank to perform
differential analysis on. Should be one of
<code>phyloseq::rank_names(phyloseq)</code>, or "all" means to summarize the taxa by
the top taxa ranks (<code>summarize_taxa(ps, level = rank_names(ps)[1])</code>), or
"none" means perform differential analysis on the original taxa
(<code>taxa_names(phyloseq)</code>, e.g., OTU or ASV).</p></dd>
<dt>transform</dt>
<dd><p>character, the methods used to transform the microbial
abundance. See <code><a href="transform_abundances.html">transform_abundances()</a></code> for more details. The
options include:</p><ul><li><p>"identity", return the original data without any transformation
(default).</p></li>
<li><p>"log10", the transformation is <code>log10(object)</code>, and if the data contains
zeros the transformation is <code>log10(1 + object)</code>.</p></li>
<li><p>"log10p", the transformation is <code>log10(1 + object)</code>.</p></li>
</ul></dd>
<dt>norm</dt>
<dd><p>the methods used to normalize the microbial abundance data. See
<code><a href="normalize-methods.html">normalize()</a></code> for more details.
Options include:</p><ul><li><p>"none": do not normalize.</p></li>
<li><p>"rarefy": random subsampling counts to the smallest library size in the
data set.</p></li>
<li><p>"TSS": total sum scaling, also referred to as "relative abundance", the
abundances were normalized by dividing the corresponding sample library
size.</p></li>
<li><p>"TMM": trimmed mean of m-values. First, a sample is chosen as reference.
The scaling factor is then derived using a weighted trimmed mean over the
differences of the log-transformed gene-count fold-change between the
sample and the reference.</p></li>
<li><p>"RLE", relative log expression, RLE uses a pseudo-reference calculated
using the geometric mean of the gene-specific abundances over all
samples. The scaling factors are then calculated as the median of the
gene counts ratios between the samples and the reference.</p></li>
<li><p>"CSS": cumulative sum scaling, calculates scaling factors as the
cumulative sum of gene abundances up to a data-derived threshold.</p></li>
<li><p>"CLR": centered log-ratio normalization.</p></li>
<li><p>"CPM": pre-sample normalization of the sum of the values to 1e+06.</p></li>
</ul></dd>
<dt>norm_para</dt>
<dd><p>named <code>list</code>. other arguments passed to specific
normalization methods.  Most users will not need to pass any additional
arguments here.</p></dd>
<dt>nfolds</dt>
<dd><p>the number of splits in CV.</p></dd>
<dt>nrepeats</dt>
<dd><p>the number of complete sets of folds to compute.</p></dd>
<dt>sampling</dt>
<dd><p>a single character value describing the type of additional
sampling that is conducted after resampling (usually to resolve class
imbalances). Values are "none", "down", "up", "smote", or "rose". For
more details see <code><a href="https://rdrr.io/pkg/caret/man/trainControl.html" class="external-link">caret::trainControl()</a></code>.</p></dd>
<dt>tune_length</dt>
<dd><p>an integer denoting the amount of granularity in the
tuning parameter grid. For more details see <code><a href="https://rdrr.io/pkg/caret/man/train.html" class="external-link">caret::train()</a></code>.</p></dd>
<dt>top_n</dt>
<dd><p>an integer denoting the top <code>n</code> features as the biomarker
according the importance score.</p></dd>
<dt>method</dt>
<dd><p>supervised learning method, options are "LR" (logistic
regression), "RF" (rando forest), or "SVM" (support vector machine).</p></dd>
<dt>...</dt>
<dd><p>extra arguments passed to the classification. e.g., <code>importance</code>
for <code>randomForest::randomForest</code>.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>a <a href="microbiomeMarker-class.html">microbiomeMarker</a> object.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Only support two groups comparison in the current version. And the
marker was selected based on its importance score. Moreover, The
hyper-parameters are selected automatically by a grid-search based method
in the N-time K-fold cross-validation. Thus, the identified biomarker based
can be biased due to model overfitting for small datasets (e.g., with less
than 100 samples).</p>
<p>The argument <code>top_n</code> is used to denote the number of markers based on the
importance score. There is no rule or principle on how to select <code>top_n</code>,
however, usually it is very useful to try a different <code>top_n</code> and compare
the performance of the marker predictions for the testing data.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="https://rdrr.io/pkg/caret/man/train.html" class="external-link">caret::train()</a></code>,<code><a href="https://rdrr.io/pkg/caret/man/trainControl.html" class="external-link">caret::trainControl()</a></code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Yang Cao</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">enterotypes_arumugam</span><span class="op">)</span></span>
<span class="r-in"><span class="co"># small example phyloseq object for test</span></span>
<span class="r-in"><span class="va">ps_small</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_taxa-methods.html" class="external-link">subset_taxa</a></span><span class="op">(</span></span>
<span class="r-in">    <span class="va">enterotypes_arumugam</span>,</span>
<span class="r-in">    <span class="va">Phylum</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Firmicutes"</span>, <span class="st">"Bacteroidetes"</span><span class="op">)</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2021</span><span class="op">)</span></span>
<span class="r-in"><span class="va">mm</span> <span class="op">&lt;-</span> <span class="fu">run_sl</span><span class="op">(</span></span>
<span class="r-in">    <span class="va">ps_small</span>,</span>
<span class="r-in">    group <span class="op">=</span> <span class="st">"Gender"</span>,</span>
<span class="r-in">    taxa_rank <span class="op">=</span> <span class="st">"Genus"</span>,</span>
<span class="r-in">    nfolds <span class="op">=</span> <span class="fl">2</span>,</span>
<span class="r-in">    nrepeats <span class="op">=</span> <span class="fl">1</span>,</span>
<span class="r-in">    top_n <span class="op">=</span> <span class="fl">15</span>,</span>
<span class="r-in">    norm <span class="op">=</span> <span class="st">"TSS"</span>,</span>
<span class="r-in">    method <span class="op">=</span> <span class="st">"LR"</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: ggplot2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: lattice</span>
<span class="r-in"><span class="va">mm</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> microbiomeMarker-class inherited from phyloseq-class</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> normalization method:              [ TSS ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> microbiome marker identity method: [ logistic regression ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> marker_table() Marker Table:       [ 15 microbiome markers with 3 variables ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> otu_table()    OTU Table:          [ 91 taxa and  39 samples ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> sample_data()  Sample Data:        [ 39 samples by  9 sample variables ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> tax_table()    Taxonomy Table:     [ 91 taxa by 1 taxonomic ranks ]</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Hua Zou.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer></div>

  


  

  </body></html>

