<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ MicrobiomeAnalysis</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MicrobiomeAnalysis</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Introduction.html">MicrobiomeAnalysis: A comprehensive R package for managing and analyzing microbiome data</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/HuaZou/MicrobiomeAnalysis/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="data-import">Data import <a href="#data-import" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions for importing external data and converting other R object as phyloseq or reverse converting</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="import_SE.html">import_SE()</a></code> </p>
        </td>
        <td><p>Convert inputs into SummarizedExperiment object</p></td>
      </tr><tr><td>
          <p><code><a href="import_dada2.html">import_dada2()</a></code> </p>
        </td>
        <td><p>Convert the output of dada2 into phyloseq object</p></td>
      </tr><tr><td>
          <p><code><a href="import_qiime2.html">import_qiime2()</a></code> </p>
        </td>
        <td><p>Convert the qiime2 output of dada2 into phyloseq object</p></td>
      </tr><tr><td>
          <p><code><a href="phyloseq2DESeq2.html">phyloseq2DESeq2()</a></code> </p>
        </td>
        <td><p>Convert <code>phyloseq-class</code> object to <code>DESeqDataSet-class</code> object</p></td>
      </tr><tr><td>
          <p><code><a href="phyloseq2edgeR.html">phyloseq2edgeR()</a></code> </p>
        </td>
        <td><p>Convert phyloseq data to edgeR <code>DGEList</code> object</p></td>
      </tr><tr><td>
          <p><code><a href="phyloseq2metagenomeSeq.html">phyloseq2metagenomeSeq()</a></code> <code><a href="phyloseq2metagenomeSeq.html">otu_table2metagenomeSeq()</a></code> </p>
        </td>
        <td><p>Convert phyloseq data to MetagenomeSeq <code>MRexperiment</code> object</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="miscellaneous">Miscellaneous <a href="#miscellaneous" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to manipulate object</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="aggregate_taxa.html">aggregate_taxa()</a></code> </p>
        </td>
        <td><p>Aggregate Taxa</p></td>
      </tr><tr><td>
          <p><code><a href="summarize_taxa.html">summarize_taxa()</a></code> </p>
        </td>
        <td><p>Summarize taxa into a taxonomic level within each sample</p></td>
      </tr><tr><td>
          <p><code><a href="subset_marker.html">subset_marker()</a></code> </p>
        </td>
        <td><p>Subset microbiome markers</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="data-processing">Data Processing <a href="#data-processing" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to precess data before test</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="abundances-methods.html">abundances()</a></code> </p>
        </td>
        <td><p>Extract taxa abundances</p></td>
      </tr><tr><td>
          <p><code><a href="transform_abundances.html">transform_abundances()</a></code> </p>
        </td>
        <td><p>Transform the abundances in <code>profile</code> sample by sample</p></td>
      </tr><tr><td>
          <p><code><a href="trim_prevalence.html">trim_prevalence()</a></code> </p>
        </td>
        <td><p>Trimming samples or features whose prevalence is less than threshold</p></td>
      </tr><tr><td>
          <p><code><a href="filter_abundance.html">filter_abundance()</a></code> </p>
        </td>
        <td><p>Filtering feature who is low relative abundance or unclassified</p></td>
      </tr><tr><td>
          <p><code><a href="impute_abundance.html">impute_abundance()</a></code> </p>
        </td>
        <td><p>Imputation methods on missing value</p></td>
      </tr><tr><td>
          <p><code><a href="normalize-methods.html">normalize(<i>&lt;phyloseq&gt;</i>)</a></code> <code><a href="normalize-methods.html">normalize(<i>&lt;otu_table&gt;</i>)</a></code> <code><a href="normalize-methods.html">normalize(<i>&lt;data.frame&gt;</i>)</a></code> <code><a href="normalize-methods.html">normalize(<i>&lt;matrix&gt;</i>)</a></code> <code><a href="normalize-methods.html">norm_rarefy()</a></code> <code><a href="normalize-methods.html">norm_tss()</a></code> <code><a href="normalize-methods.html">norm_css()</a></code> <code><a href="normalize-methods.html">norm_rle()</a></code> <code><a href="normalize-methods.html">norm_tmm()</a></code> <code><a href="normalize-methods.html">norm_clr()</a></code> <code><a href="normalize-methods.html">norm_cpm()</a></code> </p>
        </td>
        <td><p>Normalize the microbial abundance data</p></td>
      </tr><tr><td>
          <p><code><a href="scale_variables.html">scale_variables()</a></code> </p>
        </td>
        <td><p>Data scaling adjusts each variable/feature</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="example-data">Example data <a href="#example-data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="data-Zeybel_2022_gut.html">data-Zeybel_2022_gut</a></code> <code><a href="data-Zeybel_2022_gut.html">Zeybel_2022_gut</a></code> </p>
        </td>
        <td><p>gut microbiota from Zeybel et al. - 2022 paper</p></td>
      </tr><tr><td>
          <p><code><a href="data-Zeybel_2022_gut_paired.html">data-Zeybel_2022_gut_paired</a></code> <code><a href="data-Zeybel_2022_gut_paired.html">Zeybel_2022_gut_paired</a></code> </p>
        </td>
        <td><p>Paired gut microbiota from Zeybel et al. - 2022 paper</p></td>
      </tr><tr><td>
          <p><code><a href="data-Zeybel_2022_metabolite.html">data-Zeybel_2022_metabolite</a></code> <code><a href="data-Zeybel_2022_metabolite.html">Zeybel_2022_metabolite</a></code> </p>
        </td>
        <td><p>fecal metabolites from Zeybel et al. - 2022 paper</p></td>
      </tr><tr><td>
          <p><code><a href="data-Zeybel_2022_metabolite_paired.html">data-Zeybel_2022_metabolite_paired</a></code> <code><a href="data-Zeybel_2022_metabolite_paired.html">Zeybel_2022_metabolite_paired</a></code> </p>
        </td>
        <td><p>Paired fecal metabolites from Zeybel et al. - 2022 paper</p></td>
      </tr><tr><td>
          <p><code><a href="data-Zeybel_2022_oral.html">data-Zeybel_2022_oral</a></code> <code><a href="data-Zeybel_2022_oral.html">Zeybel_2022_oral</a></code> </p>
        </td>
        <td><p>oral microbiota from Zeybel et al. - 2022 paper</p></td>
      </tr><tr><td>
          <p><code><a href="data-Zeybel_2022_oral_paired.html">data-Zeybel_2022_oral_paired</a></code> <code><a href="data-Zeybel_2022_oral_paired.html">Zeybel_2022_oral_paired</a></code> </p>
        </td>
        <td><p>Paired oral microbiota from Zeybel et al. - 2022 paper</p></td>
      </tr><tr><td>
          <p><code><a href="data-Zeybel_2022_protein.html">data-Zeybel_2022_protein</a></code> <code><a href="data-Zeybel_2022_protein.html">Zeybel_2022_protein</a></code> </p>
        </td>
        <td><p>serum protein from Zeybel et al. - 2022 paper</p></td>
      </tr><tr><td>
          <p><code><a href="data-Zeybel_2022_protein_paired.html">data-Zeybel_2022_protein_paired</a></code> <code><a href="data-Zeybel_2022_protein_paired.html">Zeybel_2022_protein_paired</a></code> </p>
        </td>
        <td><p>Paired serum protein from Zeybel et al. - 2022 paper</p></td>
      </tr><tr><td>
          <p><code><a href="data-caporaso.html">data-caporaso</a></code> <code><a href="data-caporaso.html">caporaso</a></code> </p>
        </td>
        <td><p>16S rRNA data from "Moving pictures of the human microbiome"</p></td>
      </tr><tr><td>
          <p><code><a href="data-ecam.html">data-ecam</a></code> <code><a href="data-ecam.html">ecam</a></code> </p>
        </td>
        <td><p>Data from Early Childhood Antibiotics and the Microbiome (ECAM) study</p></td>
      </tr><tr><td>
          <p><code><a href="data-enterotypes_arumugam.html">data-enterotypes_arumugam</a></code> <code><a href="data-enterotypes_arumugam.html">enterotypes_arumugam</a></code> </p>
        </td>
        <td><p>Enterotypes data of 39 samples</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="community-analysis">Community Analysis <a href="#community-analysis" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to investigate association between variables and microbiota community</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="run_betadisper.html">run_betadisper()</a></code> </p>
        </td>
        <td><p>Multivariate homogeneity of groups dispersions (variances)</p></td>
      </tr><tr><td>
          <p><code><a href="run_ANOSIM.html">run_ANOSIM()</a></code> </p>
        </td>
        <td><p>Analysis of Similarity (ANOSIM)</p></td>
      </tr><tr><td>
          <p><code><a href="run_distance.html">run_distance()</a></code> </p>
        </td>
        <td><p>Calculate the distance among samples</p></td>
      </tr><tr><td>
          <p><code><a href="run_MANTEL.html">run_MANTEL()</a></code> </p>
        </td>
        <td><p>Mantel Test (MANTEL)</p></td>
      </tr><tr><td>
          <p><code><a href="run_MRPP.html">run_MRPP()</a></code> </p>
        </td>
        <td><p>Multi-response Permutation Procedures (MRPP)</p></td>
      </tr><tr><td>
          <p><code><a href="run_PERMANOVA.html">run_PERMANOVA()</a></code> </p>
        </td>
        <td><p>Permutational multivariate analysis of variance (PERMANOVA)</p></td>
      </tr><tr><td>
          <p><code><a href="get_alphaindex.html">get_alphaindex()</a></code> </p>
        </td>
        <td><p>Calculating index of Alpha diversity on microbiota data</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="differential-analysis">Differential analysis <a href="#differential-analysis" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions for identifying the microbiome markers</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="run_ANOSIM.html">run_ANOSIM()</a></code> </p>
        </td>
        <td><p>Analysis of Similarity (ANOSIM)</p></td>
      </tr><tr><td>
          <p><code><a href="run_CI.html">run_CI()</a></code> </p>
        </td>
        <td><p>Confidence Interval</p></td>
      </tr><tr><td>
          <p><code><a href="run_MANTEL.html">run_MANTEL()</a></code> </p>
        </td>
        <td><p>Mantel Test (MANTEL)</p></td>
      </tr><tr><td>
          <p><code><a href="run_MRPP.html">run_MRPP()</a></code> </p>
        </td>
        <td><p>Multi-response Permutation Procedures (MRPP)</p></td>
      </tr><tr><td>
          <p><code><a href="run_OddRatio.html">run_OddRatio()</a></code> </p>
        </td>
        <td><p>95% confidential interval Odds Ratio</p></td>
      </tr><tr><td>
          <p><code><a href="run_PERMANOVA.html">run_PERMANOVA()</a></code> </p>
        </td>
        <td><p>Permutational multivariate analysis of variance (PERMANOVA)</p></td>
      </tr><tr><td>
          <p><code><a href="run_STDERR.html">run_STDERR()</a></code> </p>
        </td>
        <td><p>Standard Error</p></td>
      </tr><tr><td>
          <p><code><a href="run_aldex.html">run_aldex()</a></code> </p>
        </td>
        <td><p>Perform differential analysis using ALDEx2</p></td>
      </tr><tr><td>
          <p><code><a href="run_ancom.html">run_ancom()</a></code> </p>
        </td>
        <td><p>Perform differential analysis using ANCOM</p></td>
      </tr><tr><td>
          <p><code><a href="run_ancombc.html">run_ancombc()</a></code> </p>
        </td>
        <td><p>Differential analysis of compositions of microbiomes with bias correction
(ANCOM-BC).</p></td>
      </tr><tr><td>
          <p><code><a href="run_betadisper.html">run_betadisper()</a></code> </p>
        </td>
        <td><p>Multivariate homogeneity of groups dispersions (variances)</p></td>
      </tr><tr><td>
          <p><code><a href="run_deseq2.html">run_deseq2()</a></code> </p>
        </td>
        <td><p>Perform DESeq differential analysis</p></td>
      </tr><tr><td>
          <p><code><a href="run_distance.html">run_distance()</a></code> </p>
        </td>
        <td><p>Calculate the distance among samples</p></td>
      </tr><tr><td>
          <p><code><a href="run_edger.html">run_edger()</a></code> </p>
        </td>
        <td><p>Perform differential analysis using edgeR</p></td>
      </tr><tr><td>
          <p><code><a href="run_group.CI.html">run_group.CI()</a></code> </p>
        </td>
        <td><p>Group Confidence Interval</p></td>
      </tr><tr><td>
          <p><code><a href="run_group.STDERR.html">run_group.STDERR()</a></code> </p>
        </td>
        <td><p>Group Standard Error Interval</p></td>
      </tr><tr><td>
          <p><code><a href="run_lefse.html">run_lefse()</a></code> </p>
        </td>
        <td><p>Liner discriminant analysis (LDA) effect size (LEFSe) analysis</p></td>
      </tr><tr><td>
          <p><code><a href="run_limma_voom.html">run_limma_voom()</a></code> </p>
        </td>
        <td><p>Differential analysis using limma-voom</p></td>
      </tr><tr><td>
          <p><code><a href="run_marker.html">run_marker()</a></code> </p>
        </td>
        <td><p>Find makers (differentially expressed metagenomic features)</p></td>
      </tr><tr><td>
          <p><code><a href="run_metagenomeseq.html">run_metagenomeseq()</a></code> </p>
        </td>
        <td><p>metagenomeSeq differential analysis</p></td>
      </tr><tr><td>
          <p><code><a href="run_simple_stat.html">run_simple_stat()</a></code> </p>
        </td>
        <td><p>Simple statistical analysis of metagenomic profiles</p></td>
      </tr><tr><td>
          <p><code><a href="run_sl.html">run_sl()</a></code> </p>
        </td>
        <td><p>Identify biomarkers using supervised leaning (SL) methods</p></td>
      </tr><tr><td>
          <p><code><a href="run_summarySE.html">run_summarySE()</a></code> </p>
        </td>
        <td><p>Summarizes data for mean median sd etc</p></td>
      </tr><tr><td>
          <p><code><a href="run_test_multiple_groups.html">run_test_multiple_groups()</a></code> </p>
        </td>
        <td><p>Statistical test for multiple groups</p></td>
      </tr><tr><td>
          <p><code><a href="run_test_two_groups.html">run_test_two_groups()</a></code> </p>
        </td>
        <td><p>Statistical test between two groups</p></td>
      </tr><tr><td>
          <p><code><a href="postHocTest.html">postHocTest()</a></code> </p>
        </td>
        <td><p>Build postHocTest object</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="microbiome-marker">Microbiome marker <a href="#microbiome-marker" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>S4 class and methods for microbiomeMarker</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="microbiomeMarker.html">microbiomeMarker()</a></code> </p>
        </td>
        <td><p>Build microbiomeMarker-class objects</p></td>
      </tr><tr><td>
          <p><code><a href="microbiomeMarker-class.html">show(<i>&lt;microbiomeMarker&gt;</i>)</a></code> </p>
        </td>
        <td><p>The main class for microbiomeMarker data</p></td>
      </tr><tr><td>
          <p><code><a href="marker_table-class.html">marker_table-class</a></code> </p>
        </td>
        <td><p>The S4 class for storing microbiome marker information</p></td>
      </tr><tr><td>
          <p><code><a href="marker_table-methods.html">marker_table()</a></code> </p>
        </td>
        <td><p>Build or access the marker_table</p></td>
      </tr><tr><td>
          <p><code><a href="assign-marker_table.html">`marker_table&lt;-`()</a></code> </p>
        </td>
        <td><p>Assign marker_table to <code>object</code></p></td>
      </tr><tr><td>
          <p><code><a href="assign-otu_table.html">`otu_table&lt;-`(<i>&lt;microbiomeMarker&gt;</i>,<i>&lt;otu_table&gt;</i>)</a></code> <code><a href="assign-otu_table.html">`otu_table&lt;-`(<i>&lt;microbiomeMarker&gt;</i>,<i>&lt;phyloseq&gt;</i>)</a></code> <code><a href="assign-otu_table.html">`otu_table&lt;-`(<i>&lt;microbiomeMarker&gt;</i>,<i>&lt;microbiomeMarker&gt;</i>)</a></code> </p>
        </td>
        <td><p>Assign a new OTU table</p></td>
      </tr><tr><td>
          <p><code><a href="abundances-methods.html">abundances()</a></code> </p>
        </td>
        <td><p>Extract taxa abundances</p></td>
      </tr><tr><td>
          <p><code><a href="nmarker-methods.html">nmarker()</a></code> </p>
        </td>
        <td><p>Get the number of microbiome markers</p></td>
      </tr><tr><td>
          <p><code><a href="postHocTest-class.html">show(<i>&lt;postHocTest&gt;</i>)</a></code> </p>
        </td>
        <td><p>The postHocTest Class, represents the result of post-hoc test result among
multiple groups</p></td>
      </tr><tr><td>
          <p><code><a href="extract-methods.html">`[`(<i>&lt;marker_table&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;ANY&gt;</i>)</a></code> </p>
        </td>
        <td><p>Extract <code>marker_table</code> object</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="statistical-tookits">Statistical tookits <a href="#statistical-tookits" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to do statistical test</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="calculate_median_abundance.html">calculate_median_abundance()</a></code> </p>
        </td>
        <td><p>Calculate the Median abundance of features per group (two groups)</p></td>
      </tr><tr><td>
          <p><code><a href="calculate_mean_abundance.html">calculate_mean_abundance()</a></code> </p>
        </td>
        <td><p>Calculate the Mean abundance of features per group (two groups)</p></td>
      </tr><tr><td>
          <p><code><a href="calculate_MeanRank_abundance.html">calculate_MeanRank_abundance()</a></code> </p>
        </td>
        <td><p>Calculate the Rank Mean abundance of features per group(two groups)</p></td>
      </tr><tr><td>
          <p><code><a href="calculate_geometricmean_abundance.html">calculate_geometricmean_abundance()</a></code> </p>
        </td>
        <td><p>Calculate the Mean abundance of features per group (two groups)</p></td>
      </tr><tr><td>
          <p><code><a href="calculate_occurrence_taxa.html">calculate_occurrence_taxa()</a></code> </p>
        </td>
        <td><p>Calculate the Occurrence of features per group (two groups)</p></td>
      </tr><tr><td>
          <p><code><a href="run_OddRatio.html">run_OddRatio()</a></code> </p>
        </td>
        <td><p>95% confidential interval Odds Ratio</p></td>
      </tr><tr><td>
          <p><code><a href="run_CI.html">run_CI()</a></code> </p>
        </td>
        <td><p>Confidence Interval</p></td>
      </tr><tr><td>
          <p><code><a href="run_group.CI.html">run_group.CI()</a></code> </p>
        </td>
        <td><p>Group Confidence Interval</p></td>
      </tr><tr><td>
          <p><code><a href="run_STDERR.html">run_STDERR()</a></code> </p>
        </td>
        <td><p>Standard Error</p></td>
      </tr><tr><td>
          <p><code><a href="run_group.STDERR.html">run_group.STDERR()</a></code> </p>
        </td>
        <td><p>Group Standard Error Interval</p></td>
      </tr><tr><td>
          <p><code><a href="run_summarySE.html">run_summarySE()</a></code> </p>
        </td>
        <td><p>Summarizes data for mean median sd etc</p></td>
      </tr><tr><td>
          <p><code><a href="insertRow.html">insertRow()</a></code> </p>
        </td>
        <td><p>Insert Row into a Matrix</p></td>
      </tr><tr><td>
          <p><code><a href="insertCol.html">insertCol()</a></code> </p>
        </td>
        <td><p>Insert Row into a Matrix</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="additional-tookits">Additional tookits <a href="#additional-tookits" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Additional Functions</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="default_color.html">default_color()</a></code> </p>
        </td>
        <td><p>generate multiple colors for visualization</p></td>
      </tr><tr><td>
          <p><code><a href="check_sample_names.html">check_sample_names()</a></code> </p>
        </td>
        <td><p>check whether samples' names start with numeric and then paste "S_"</p></td>
      </tr><tr><td>
          <p><code><a href="get_GUniFrac.html">get_GUniFrac()</a></code> </p>
        </td>
        <td><p>Generalized UniFrac distances for comparing microbial communities</p></td>
      </tr><tr><td>
          <p><code><a href="get_eigValue.html">get_eigValue()</a></code> </p>
        </td>
        <td><p>Extract and visualize the eigenvalues/variances of dimensions</p></td>
      </tr><tr><td>
          <p><code><a href="confounder.html">confounder()</a></code> </p>
        </td>
        <td><p>Confounder analysis</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Hua Zou.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

