<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>MicrobiomeAnalysis: A comprehensive R package for managing and analyzing microbiome data • MicrobiomeAnalysis</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="MicrobiomeAnalysis: A comprehensive R package for managing and analyzing microbiome data">
<meta property="og:description" content="MicrobiomeAnalysis">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MicrobiomeAnalysis</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Introduction.html">MicrobiomeAnalysis: A comprehensive R package for managing and analyzing microbiome data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/HuaZou/MicrobiomeAnalysis/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>MicrobiomeAnalysis: A comprehensive R package
for managing and analyzing microbiome data</h1>
                        <h4 data-toc-skip class="author">Hua Zou</h4>
            <address class="author_afil">
      UCAS<br><a class="author_email" href="mailto:#"></a><a href="mailto:zouhua1@outlook.com" class="email">zouhua1@outlook.com</a>
      </address>
                  
            <h4 data-toc-skip class="date">2023-01-10</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/HuaZou/MicrobiomeAnalysis/blob/HEAD/vignettes/Introduction.Rmd" class="external-link"><code>vignettes/Introduction.Rmd</code></a></small>
      <div class="hidden name"><code>Introduction.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview-of-microbiomeanalysis">Overview of <strong>MicrobiomeAnalysis</strong><a class="anchor" aria-label="anchor" href="#overview-of-microbiomeanalysis"></a>
</h2>
<div class="figure" style="text-align: center">
<img src="../reference/figures/Schematic.png" alt="The Overview of MicrobiomeAnalysis" width="80%"><p class="caption">
The Overview of MicrobiomeAnalysis
</p>
</div>
</div>
<div class="section level2">
<h2 id="convert-inputs-into-phyloseq-class-object">Convert inputs into phyloseq-class object<a class="anchor" aria-label="anchor" href="#convert-inputs-into-phyloseq-class-object"></a>
</h2>
<div class="section level3">
<h3 id="inputs-from-dada2">inputs from dada2<a class="anchor" aria-label="anchor" href="#inputs-from-dada2"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seq_tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>
  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"dada2_seqtab.rds"</span>,
     package <span class="op">=</span> <span class="st">"MicrobiomeAnalysis"</span><span class="op">)</span><span class="op">)</span>

<span class="va">tax_tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>
  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"dada2_taxtab.rds"</span>,
    package <span class="op">=</span> <span class="st">"MicrobiomeAnalysis"</span><span class="op">)</span><span class="op">)</span>

<span class="va">sam_tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span>
  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"dada2_samdata.txt"</span>,
     package <span class="op">=</span> <span class="st">"MicrobiomeAnalysis"</span><span class="op">)</span>, 
  sep <span class="op">=</span> <span class="st">"\t"</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>

<span class="va">ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_dada2.html">import_dada2</a></span><span class="op">(</span>seq_tab <span class="op">=</span> <span class="va">seq_tab</span>, 
                   tax_tab <span class="op">=</span> <span class="va">tax_tab</span>, 
                   sam_tab <span class="op">=</span> <span class="va">sam_tab</span><span class="op">)</span>

<span class="va">ps</span></code></pre></div>
<pre><code><span class="co">## phyloseq-class experiment-level object</span>
<span class="co">## otu_table()   OTU Table:         [ 232 taxa and 20 samples ]</span>
<span class="co">## sample_data() Sample Data:       [ 20 samples by 4 sample variables ]</span>
<span class="co">## tax_table()   Taxonomy Table:    [ 232 taxa by 6 taxonomic ranks ]</span>
<span class="co">## refseq()      DNAStringSet:      [ 232 reference sequences ]</span></code></pre>
</div>
<div class="section level3">
<h3 id="inputs-from-qiime2">inputs from qiime2<a class="anchor" aria-label="anchor" href="#inputs-from-qiime2"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">otuqza_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>
     <span class="st">"extdata"</span>, <span class="st">"table.qza"</span>,
     package <span class="op">=</span> <span class="st">"MicrobiomeAnalysis"</span><span class="op">)</span>

<span class="va">taxaqza_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>
     <span class="st">"extdata"</span>, <span class="st">"taxonomy.qza"</span>,
     package <span class="op">=</span> <span class="st">"MicrobiomeAnalysis"</span><span class="op">)</span>

<span class="va">sample_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>
     <span class="st">"extdata"</span>, <span class="st">"sample-metadata.tsv"</span>,
     package <span class="op">=</span> <span class="st">"MicrobiomeAnalysis"</span><span class="op">)</span>

<span class="va">treeqza_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>
     <span class="st">"extdata"</span>, <span class="st">"tree.qza"</span>,
     package <span class="op">=</span> <span class="st">"MicrobiomeAnalysis"</span><span class="op">)</span>
 
<span class="va">ps_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_qiime2.html">import_qiime2</a></span><span class="op">(</span>
     otu_qza <span class="op">=</span> <span class="va">otuqza_file</span>, 
     taxa_qza <span class="op">=</span> <span class="va">taxaqza_file</span>,
     sam_tab <span class="op">=</span> <span class="va">sample_file</span>, 
     tree_qza <span class="op">=</span> <span class="va">treeqza_file</span><span class="op">)</span>
<span class="va">ps_object</span></code></pre></div>
<pre><code><span class="co">## phyloseq-class experiment-level object</span>
<span class="co">## otu_table()   OTU Table:         [ 770 taxa and 34 samples ]</span>
<span class="co">## sample_data() Sample Data:       [ 34 samples by 9 sample variables ]</span>
<span class="co">## tax_table()   Taxonomy Table:    [ 770 taxa by 7 taxonomic ranks ]</span>
<span class="co">## phy_tree()    Phylogenetic Tree: [ 770 tips and 768 internal nodes ]</span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="data-processing">Data Processing<a class="anchor" aria-label="anchor" href="#data-processing"></a>
</h2>
<div class="section level3">
<h3 id="extract-specific-levels-phyloseq-class-object">Extract specific levels phyloseq-class object<a class="anchor" aria-label="anchor" href="#extract-specific-levels-phyloseq-class-object"></a>
</h3>
<ul>
<li>Extracting “Genus” levels phyloseq object</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ps_genus</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregate_taxa.html">aggregate_taxa</a></span><span class="op">(</span><span class="va">ps_object</span>, 
                           level <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span>
<span class="va">ps_genus</span></code></pre></div>
<pre><code><span class="co">## phyloseq-class experiment-level object</span>
<span class="co">## otu_table()   OTU Table:         [ 277 taxa and 34 samples ]</span>
<span class="co">## sample_data() Sample Data:       [ 34 samples by 9 sample variables ]</span>
<span class="co">## tax_table()   Taxonomy Table:    [ 277 taxa by 7 taxonomic ranks ]</span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ps_genus</span><span class="op">@</span><span class="va">otu_table</span><span class="op">@</span><span class="va">.Data</span>, <span class="fl">3</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##                                  L1S105 L1S140 L1S208 L1S257 L1S281 L1S57 L1S76</span>
<span class="co">## g__Candidatus_Nitrososphaera          0      0      0      0      0     0     0</span>
<span class="co">## g__Ellin6075_unclassified             0      0      0      0      0     0     0</span>
<span class="co">## g__Acidimicrobiales_unclassified      0      0      0      0      0     0     0</span>
<span class="co">##                                  L1S8 L2S155 L2S175 L2S204 L2S222 L2S240 L2S309</span>
<span class="co">## g__Candidatus_Nitrososphaera        0      0      0      0      0      0      0</span>
<span class="co">## g__Ellin6075_unclassified           0      0      0      0      0      0      0</span>
<span class="co">## g__Acidimicrobiales_unclassified    0      0      0      0      0      0      0</span>
<span class="co">##                                  L2S357 L2S382 L3S242 L3S294 L3S313 L3S341</span>
<span class="co">## g__Candidatus_Nitrososphaera          0      0      0      0      0      0</span>
<span class="co">## g__Ellin6075_unclassified             0      0      0      0      0      0</span>
<span class="co">## g__Acidimicrobiales_unclassified      0      0      0      0      0      0</span>
<span class="co">##                                  L3S360 L3S378 L4S112 L4S137 L4S63 L5S104</span>
<span class="co">## g__Candidatus_Nitrososphaera          0      0      4      0     8      0</span>
<span class="co">## g__Ellin6075_unclassified             0      0      6      0    13      0</span>
<span class="co">## g__Acidimicrobiales_unclassified      0      0      0      0     4      0</span>
<span class="co">##                                  L5S155 L5S174 L5S203 L5S222 L5S240 L6S20 L6S68</span>
<span class="co">## g__Candidatus_Nitrososphaera          0      0      0      0      0     0     0</span>
<span class="co">## g__Ellin6075_unclassified             0      0      0      0      0     0     0</span>
<span class="co">## g__Acidimicrobiales_unclassified      0      0      0      0      0     0     0</span>
<span class="co">##                                  L6S93</span>
<span class="co">## g__Candidatus_Nitrososphaera         0</span>
<span class="co">## g__Ellin6075_unclassified            0</span>
<span class="co">## g__Acidimicrobiales_unclassified     0</span></code></pre>
</div>
<div class="section level3">
<h3 id="summarize-the-specific-taxonomic-levels">Summarize the specific taxonomic levels<a class="anchor" aria-label="anchor" href="#summarize-the-specific-taxonomic-levels"></a>
</h3>
<ul>
<li>Phyloseq object contains from Kingdom to the the specific taxonomic
levels (Phylum)</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ps_summarize_genus</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_taxa.html">summarize_taxa</a></span><span class="op">(</span>
    <span class="va">ps_object</span>, 
    level <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span>
<span class="va">ps_summarize_genus</span></code></pre></div>
<pre><code><span class="co">## phyloseq-class experiment-level object</span>
<span class="co">## otu_table()   OTU Table:         [ 628 taxa and 34 samples ]</span>
<span class="co">## sample_data() Sample Data:       [ 34 samples by 9 sample variables ]</span>
<span class="co">## tax_table()   Taxonomy Table:    [ 628 taxa by 1 taxonomic ranks ]</span></code></pre>
</div>
<div class="section level3">
<h3 id="data-transformation">Data Transformation<a class="anchor" aria-label="anchor" href="#data-transformation"></a>
</h3>
<ul>
<li>Transform the taxa abundances in <code>otu_table</code>, which
applying a mathematical transformation on individual values
themselves.</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ps_genus_transform</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/transform_abundances.html">transform_abundances</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">ps_genus</span>,
    transform <span class="op">=</span> <span class="st">"log10p"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ps_genus_transform</span><span class="op">@</span><span class="va">otu_table</span><span class="op">@</span><span class="va">.Data</span>, <span class="fl">3</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##                                  L1S105 L1S140 L1S208 L1S257 L1S281 L1S57 L1S76</span>
<span class="co">## g__Candidatus_Nitrososphaera          0      0      0      0      0     0     0</span>
<span class="co">## g__Ellin6075_unclassified             0      0      0      0      0     0     0</span>
<span class="co">## g__Acidimicrobiales_unclassified      0      0      0      0      0     0     0</span>
<span class="co">##                                  L1S8 L2S155 L2S175 L2S204 L2S222 L2S240 L2S309</span>
<span class="co">## g__Candidatus_Nitrososphaera        0      0      0      0      0      0      0</span>
<span class="co">## g__Ellin6075_unclassified           0      0      0      0      0      0      0</span>
<span class="co">## g__Acidimicrobiales_unclassified    0      0      0      0      0      0      0</span>
<span class="co">##                                  L2S357 L2S382 L3S242 L3S294 L3S313 L3S341</span>
<span class="co">## g__Candidatus_Nitrososphaera          0      0      0      0      0      0</span>
<span class="co">## g__Ellin6075_unclassified             0      0      0      0      0      0</span>
<span class="co">## g__Acidimicrobiales_unclassified      0      0      0      0      0      0</span>
<span class="co">##                                  L3S360 L3S378    L4S112 L4S137    L4S63 L5S104</span>
<span class="co">## g__Candidatus_Nitrososphaera          0      0 0.6020600      0 0.903090      0</span>
<span class="co">## g__Ellin6075_unclassified             0      0 0.7781513      0 1.113943      0</span>
<span class="co">## g__Acidimicrobiales_unclassified      0      0 0.0000000      0 0.602060      0</span>
<span class="co">##                                  L5S155 L5S174 L5S203 L5S222 L5S240 L6S20 L6S68</span>
<span class="co">## g__Candidatus_Nitrososphaera          0      0      0      0      0     0     0</span>
<span class="co">## g__Ellin6075_unclassified             0      0      0      0      0     0     0</span>
<span class="co">## g__Acidimicrobiales_unclassified      0      0      0      0      0     0     0</span>
<span class="co">##                                  L6S93</span>
<span class="co">## g__Candidatus_Nitrososphaera         0</span>
<span class="co">## g__Ellin6075_unclassified            0</span>
<span class="co">## g__Acidimicrobiales_unclassified     0</span></code></pre>
</div>
<div class="section level3">
<h3 id="data-imputation">Data Imputation<a class="anchor" aria-label="anchor" href="#data-imputation"></a>
</h3>
<ul>
<li>Imputing missing value of OTU_table in phyloseq-class object</li>
</ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ps_genus_impute</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/impute_abundance.html">impute_abundance</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">ps_genus</span>,
    group_name <span class="op">=</span> <span class="st">"body.site"</span>,
    ZerosAsNA <span class="op">=</span> <span class="cn">TRUE</span>,
    RemoveNA <span class="op">=</span> <span class="cn">TRUE</span>,
    cutoff <span class="op">=</span> <span class="fl">20</span>,    
    method <span class="op">=</span> <span class="st">"half_min"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ps_genus_impute</span><span class="op">@</span><span class="va">otu_table</span><span class="op">@</span><span class="va">.Data</span>, <span class="fl">3</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##                      L1S105 L1S140 L1S208 L1S257 L1S281 L1S57 L1S76 L1S8 L2S155</span>
<span class="co">## g__Corynebacterium      1.5      7    1.5    1.5    1.5   1.5   1.5  1.5   1156</span>
<span class="co">## g__Rothia               2.0      2    2.0    2.0    2.0   2.0   2.0  2.0     11</span>
<span class="co">## g__Propionibacterium    1.0      1    1.0    1.0    1.0   1.0   1.0  1.0      7</span>
<span class="co">##                      L2S175 L2S204 L2S222 L2S240 L2S309 L2S357 L2S382 L3S242</span>
<span class="co">## g__Corynebacterium     1367    106    164     29    152    345    285    1.5</span>
<span class="co">## g__Rothia                52     16      4    178     12     22     37    2.0</span>
<span class="co">## g__Propionibacterium     31      3      7      1     31     18     35    1.0</span>
<span class="co">##                      L3S294 L3S313 L3S341 L3S360 L3S378 L4S112 L4S137 L4S63</span>
<span class="co">## g__Corynebacterium      424    340     43     51    1.5   1007    943  1109</span>
<span class="co">## g__Rothia                 9     19      4      2    2.0     82     82    48</span>
<span class="co">## g__Propionibacterium      7     11      2      1    1.0     87     25   106</span>
<span class="co">##                      L5S104 L5S155 L5S174 L5S203 L5S222 L5S240 L6S20 L6S68</span>
<span class="co">## g__Corynebacterium      1.5    1.5    1.5      3    1.5    1.5   1.5   1.5</span>
<span class="co">## g__Rothia             178.0   64.0   45.0     83   31.0  104.0 108.0 224.0</span>
<span class="co">## g__Propionibacterium    1.0    1.0    1.0      1    1.0    1.0   1.0   1.0</span>
<span class="co">##                      L6S93</span>
<span class="co">## g__Corynebacterium     1.5</span>
<span class="co">## g__Rothia             38.0</span>
<span class="co">## g__Propionibacterium   1.0</span></code></pre>
</div>
<div class="section level3">
<h3 id="data-normalization">Data Normalization<a class="anchor" aria-label="anchor" href="#data-normalization"></a>
</h3>
<ul>
<li>Normalizing the OTU_table in phyloseq-class object sample by sample
to reduce the effects of systematic differences such as library
size.</li>
</ul>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ps_genus_norm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalize-methods.html">normalize</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">ps_genus</span>,
    method <span class="op">=</span> <span class="st">"TSS"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ps_genus_norm</span><span class="op">@</span><span class="va">otu_table</span><span class="op">@</span><span class="va">.Data</span>, <span class="fl">3</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##                                  L1S105 L1S140 L1S208 L1S257 L1S281 L1S57 L1S76</span>
<span class="co">## g__Candidatus_Nitrososphaera          0      0      0      0      0     0     0</span>
<span class="co">## g__Ellin6075_unclassified             0      0      0      0      0     0     0</span>
<span class="co">## g__Acidimicrobiales_unclassified      0      0      0      0      0     0     0</span>
<span class="co">##                                  L1S8 L2S155 L2S175 L2S204 L2S222 L2S240 L2S309</span>
<span class="co">## g__Candidatus_Nitrososphaera        0      0      0      0      0      0      0</span>
<span class="co">## g__Ellin6075_unclassified           0      0      0      0      0      0      0</span>
<span class="co">## g__Acidimicrobiales_unclassified    0      0      0      0      0      0      0</span>
<span class="co">##                                  L2S357 L2S382 L3S242 L3S294 L3S313 L3S341</span>
<span class="co">## g__Candidatus_Nitrososphaera          0      0      0      0      0      0</span>
<span class="co">## g__Ellin6075_unclassified             0      0      0      0      0      0</span>
<span class="co">## g__Acidimicrobiales_unclassified      0      0      0      0      0      0</span>
<span class="co">##                                  L3S360 L3S378       L4S112 L4S137        L4S63</span>
<span class="co">## g__Candidatus_Nitrososphaera          0      0 0.0004796163      0 0.0008210181</span>
<span class="co">## g__Ellin6075_unclassified             0      0 0.0007194245      0 0.0013341544</span>
<span class="co">## g__Acidimicrobiales_unclassified      0      0 0.0000000000      0 0.0004105090</span>
<span class="co">##                                  L5S104 L5S155 L5S174 L5S203 L5S222 L5S240</span>
<span class="co">## g__Candidatus_Nitrososphaera          0      0      0      0      0      0</span>
<span class="co">## g__Ellin6075_unclassified             0      0      0      0      0      0</span>
<span class="co">## g__Acidimicrobiales_unclassified      0      0      0      0      0      0</span>
<span class="co">##                                  L6S20 L6S68 L6S93</span>
<span class="co">## g__Candidatus_Nitrososphaera         0     0     0</span>
<span class="co">## g__Ellin6075_unclassified            0     0     0</span>
<span class="co">## g__Acidimicrobiales_unclassified     0     0     0</span></code></pre>
</div>
<div class="section level3">
<h3 id="data-scaling">Data Scaling<a class="anchor" aria-label="anchor" href="#data-scaling"></a>
</h3>
<ul>
<li>Data scaling adjusts each variable/feature by a scaling factor
computed based on the dispersion of the variable.</li>
</ul>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ps_genus_scale</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scale_variables.html">scale_variables</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">ps_genus</span>,
    method <span class="op">=</span> <span class="st">"zscore"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ps_genus_scale</span><span class="op">@</span><span class="va">otu_table</span><span class="op">@</span><span class="va">.Data</span>, <span class="fl">3</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##                                      L1S105     L1S140     L1S208     L1S257</span>
<span class="co">## g__Candidatus_Nitrososphaera     -0.2329302 -0.2329302 -0.2329302 -0.2329302</span>
<span class="co">## g__Ellin6075_unclassified        -0.2302519 -0.2302519 -0.2302519 -0.2302519</span>
<span class="co">## g__Acidimicrobiales_unclassified -0.1714986 -0.1714986 -0.1714986 -0.1714986</span>
<span class="co">##                                      L1S281      L1S57      L1S76       L1S8</span>
<span class="co">## g__Candidatus_Nitrososphaera     -0.2329302 -0.2329302 -0.2329302 -0.2329302</span>
<span class="co">## g__Ellin6075_unclassified        -0.2302519 -0.2302519 -0.2302519 -0.2302519</span>
<span class="co">## g__Acidimicrobiales_unclassified -0.1714986 -0.1714986 -0.1714986 -0.1714986</span>
<span class="co">##                                      L2S155     L2S175     L2S204     L2S222</span>
<span class="co">## g__Candidatus_Nitrososphaera     -0.2329302 -0.2329302 -0.2329302 -0.2329302</span>
<span class="co">## g__Ellin6075_unclassified        -0.2302519 -0.2302519 -0.2302519 -0.2302519</span>
<span class="co">## g__Acidimicrobiales_unclassified -0.1714986 -0.1714986 -0.1714986 -0.1714986</span>
<span class="co">##                                      L2S240     L2S309     L2S357     L2S382</span>
<span class="co">## g__Candidatus_Nitrososphaera     -0.2329302 -0.2329302 -0.2329302 -0.2329302</span>
<span class="co">## g__Ellin6075_unclassified        -0.2302519 -0.2302519 -0.2302519 -0.2302519</span>
<span class="co">## g__Acidimicrobiales_unclassified -0.1714986 -0.1714986 -0.1714986 -0.1714986</span>
<span class="co">##                                      L3S242     L3S294     L3S313     L3S341</span>
<span class="co">## g__Candidatus_Nitrososphaera     -0.2329302 -0.2329302 -0.2329302 -0.2329302</span>
<span class="co">## g__Ellin6075_unclassified        -0.2302519 -0.2302519 -0.2302519 -0.2302519</span>
<span class="co">## g__Acidimicrobiales_unclassified -0.1714986 -0.1714986 -0.1714986 -0.1714986</span>
<span class="co">##                                      L3S360     L3S378     L4S112     L4S137</span>
<span class="co">## g__Candidatus_Nitrososphaera     -0.2329302 -0.2329302  2.4069456 -0.2329302</span>
<span class="co">## g__Ellin6075_unclassified        -0.2302519 -0.2302519  2.2419265 -0.2302519</span>
<span class="co">## g__Acidimicrobiales_unclassified -0.1714986 -0.1714986 -0.1714986 -0.1714986</span>
<span class="co">##                                     L4S63     L5S104     L5S155     L5S174</span>
<span class="co">## g__Candidatus_Nitrososphaera     5.046821 -0.2329302 -0.2329302 -0.2329302</span>
<span class="co">## g__Ellin6075_unclassified        5.126135 -0.2302519 -0.2302519 -0.2302519</span>
<span class="co">## g__Acidimicrobiales_unclassified 5.659453 -0.1714986 -0.1714986 -0.1714986</span>
<span class="co">##                                      L5S203     L5S222     L5S240      L6S20</span>
<span class="co">## g__Candidatus_Nitrososphaera     -0.2329302 -0.2329302 -0.2329302 -0.2329302</span>
<span class="co">## g__Ellin6075_unclassified        -0.2302519 -0.2302519 -0.2302519 -0.2302519</span>
<span class="co">## g__Acidimicrobiales_unclassified -0.1714986 -0.1714986 -0.1714986 -0.1714986</span>
<span class="co">##                                       L6S68      L6S93</span>
<span class="co">## g__Candidatus_Nitrososphaera     -0.2329302 -0.2329302</span>
<span class="co">## g__Ellin6075_unclassified        -0.2302519 -0.2302519</span>
<span class="co">## g__Acidimicrobiales_unclassified -0.1714986 -0.1714986</span></code></pre>
</div>
<div class="section level3">
<h3 id="data-trimming">Data Trimming<a class="anchor" aria-label="anchor" href="#data-trimming"></a>
</h3>
<ul>
<li>Trimming samples or taxa whose prevalence is less than
threshold</li>
</ul>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ps_genus_trim</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/trim_prevalence.html">trim_prevalence</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">ps_genus</span>,
    level <span class="op">=</span> <span class="cn">NULL</span>,
    cutoff <span class="op">=</span> <span class="fl">0.1</span>,
    trim <span class="op">=</span> <span class="st">"feature"</span><span class="op">)</span>

<span class="va">ps_genus_trim</span></code></pre></div>
<pre><code><span class="co">## phyloseq-class experiment-level object</span>
<span class="co">## otu_table()   OTU Table:         [ 114 taxa and 34 samples ]</span>
<span class="co">## sample_data() Sample Data:       [ 34 samples by 9 sample variables ]</span>
<span class="co">## tax_table()   Taxonomy Table:    [ 114 taxa by 7 taxonomic ranks ]</span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ps_genus_trim</span><span class="op">@</span><span class="va">otu_table</span><span class="op">@</span><span class="va">.Data</span>, <span class="fl">3</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##                    L1S105 L1S140 L1S208 L1S257 L1S281 L1S57 L1S76 L1S8 L2S155</span>
<span class="co">## g__Actinomyces          0      0      0      0      0     0     0    0      0</span>
<span class="co">## g__Corynebacterium      0      7      0      0      0     0     0    0   1156</span>
<span class="co">## g__Rothia               0      0      0      0      0     0     0    0     11</span>
<span class="co">##                    L2S175 L2S204 L2S222 L2S240 L2S309 L2S357 L2S382 L3S242</span>
<span class="co">## g__Actinomyces          0      0      0     27      0      0      7      0</span>
<span class="co">## g__Corynebacterium   1367    106    164     29    152    345    285      0</span>
<span class="co">## g__Rothia              52     16      4    178     12     22     37      0</span>
<span class="co">##                    L3S294 L3S313 L3S341 L3S360 L3S378 L4S112 L4S137 L4S63</span>
<span class="co">## g__Actinomyces          0      0      0      0      0     13     30     7</span>
<span class="co">## g__Corynebacterium    424    340     43     51      0   1007    943  1109</span>
<span class="co">## g__Rothia               9     19      4      0      0     82     82    48</span>
<span class="co">##                    L5S104 L5S155 L5S174 L5S203 L5S222 L5S240 L6S20 L6S68 L6S93</span>
<span class="co">## g__Actinomyces         28      0     37     21     13      0    92    34    14</span>
<span class="co">## g__Corynebacterium      0      0      0      3      0      0     0     0     0</span>
<span class="co">## g__Rothia             178     64     45     83     31    104   108   224    38</span></code></pre>
</div>
<div class="section level3">
<h3 id="data-filtering">Data Filtering<a class="anchor" aria-label="anchor" href="#data-filtering"></a>
</h3>
<ul>
<li>
<p>Filtering taxa who is low relative abundance or unclassified
(Ref: <span class="citation">(Thingholm et al. 2019)</span>)</p>
<ul>
<li><p>Taxa more than Mean relative abundance across all samples:
100;</p></li>
<li><p>Taxa more than Minimum relative abundance at least one sample:
1000.</p></li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ps_genus_filter</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_abundance.html">filter_abundance</a></span><span class="op">(</span>
    object <span class="op">=</span> <span class="va">ps_genus</span>,
    level <span class="op">=</span> <span class="cn">NULL</span>,
    cutoff_mean <span class="op">=</span> <span class="fl">100</span>,
    cutoff_one <span class="op">=</span> <span class="fl">1000</span>,
    unclass <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="va">ps_genus_filter</span></code></pre></div>
<pre><code><span class="co">## phyloseq-class experiment-level object</span>
<span class="co">## otu_table()   OTU Table:         [ 7 taxa and 34 samples ]</span>
<span class="co">## sample_data() Sample Data:       [ 34 samples by 9 sample variables ]</span>
<span class="co">## tax_table()   Taxonomy Table:    [ 7 taxa by 7 taxonomic ranks ]</span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ps_genus_filter</span><span class="op">@</span><span class="va">otu_table</span><span class="op">@</span><span class="va">.Data</span>, <span class="fl">3</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##                    L1S105 L1S140 L1S208 L1S257 L1S281 L1S57 L1S76 L1S8 L2S155</span>
<span class="co">## g__Bacteroides       4521   4613   4569   3016   3078  4555  4824 4568     10</span>
<span class="co">## g__Corynebacterium      0      7      0      0      0     0     0    0   1156</span>
<span class="co">## g__Haemophilus          0      0      5      0      0     0     0    5     37</span>
<span class="co">##                    L2S175 L2S204 L2S222 L2S240 L2S309 L2S357 L2S382 L3S242</span>
<span class="co">## g__Bacteroides         10     31     52      3     32     14      0    625</span>
<span class="co">## g__Corynebacterium   1367    106    164     29    152    345    285      0</span>
<span class="co">## g__Haemophilus         95     20     15   1181     45     98    119      5</span>
<span class="co">##                    L3S294 L3S313 L3S341 L3S360 L3S378 L4S112 L4S137 L4S63</span>
<span class="co">## g__Bacteroides          0      6     11     13    763     64      5    24</span>
<span class="co">## g__Corynebacterium    424    340     43     51      0   1007    943  1109</span>
<span class="co">## g__Haemophilus         14     16     56      0      0    311    681   125</span>
<span class="co">##                    L5S104 L5S155 L5S174 L5S203 L5S222 L5S240 L6S20 L6S68 L6S93</span>
<span class="co">## g__Bacteroides         10      9      0      5      0     10     8     5     9</span>
<span class="co">## g__Corynebacterium      0      0      0      3      0      0     0     0     0</span>
<span class="co">## g__Haemophilus        529    453    254    357    390    532   940  1291   724</span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="multivariate-community-analysis">Multivariate Community Analysis<a class="anchor" aria-label="anchor" href="#multivariate-community-analysis"></a>
</h2>
<p>Know more details of the aftermentioned statistical methods to see
<span class="citation">(Xia et al. 2018)</span>.</p>
<div class="section level3">
<h3 id="multivariate-homogeneity-of-groups-dispersions-variances">Multivariate homogeneity of groups dispersions (variances)<a class="anchor" aria-label="anchor" href="#multivariate-homogeneity-of-groups-dispersions-variances"></a>
</h3>
<p><span class="citation">(Anderson, Ellingsen, and McArdle
2006)</span></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/run_betadisper.html">run_betadisper</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">ps_object</span>,
               level <span class="op">=</span> <span class="st">"Genus"</span>,
               variable <span class="op">=</span> <span class="st">"body.site"</span>, 
               method <span class="op">=</span> <span class="st">"bray"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">## Permutation test for homogeneity of multivariate dispersions</span>
<span class="co">## Permutation: free</span>
<span class="co">## Number of permutations: 999</span>
<span class="co">## </span>
<span class="co">## Response: Distances</span>
<span class="co">##           Df  Sum Sq  Mean Sq     F N.Perm Pr(&gt;F)   </span>
<span class="co">## Groups     3 0.30323 0.101077 13.08    999  0.002 **</span>
<span class="co">## Residuals 30 0.23184 0.007728                       </span>
<span class="co">## ---</span>
<span class="co">## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span>
<span class="co">## </span>
<span class="co">## Pairwise comparisons:</span>
<span class="co">## (Observed p-value below diagonal, permuted p-value above diagonal)</span>
<span class="co">##                   gut  left palm right palm tongue</span>
<span class="co">## gut                   2.0000e-02 1.0000e-03  0.341</span>
<span class="co">## left palm  1.4486e-02            9.3000e-02  0.013</span>
<span class="co">## right palm 1.6126e-05 8.7427e-02             0.001</span>
<span class="co">## tongue     3.3527e-01 9.9835e-03 6.9602e-05</span></code></pre>
</div>
<div class="section level3">
<h3 id="permutational-multivariate-analysis-of-variance-permanova">Permutational Multivariate Analysis of Variance (PERMANOVA)<a class="anchor" aria-label="anchor" href="#permutational-multivariate-analysis-of-variance-permanova"></a>
</h3>
<p><span class="citation">(Anderson 2014)</span></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/run_PERMANOVA.html">run_PERMANOVA</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">ps_object</span>,
              level <span class="op">=</span> <span class="cn">NULL</span>,
              method <span class="op">=</span> <span class="st">"bray"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##                             SumsOfSample Df  SumsOfSqs   MeanSqs   F.Model</span>
<span class="co">## sample.id                             34 33 12.9821194 0.3933976 0.0000000</span>
<span class="co">## barcode.sequence                      34 33 12.9821194 0.3933976 0.0000000</span>
<span class="co">## body.site                             34  3  5.1920140 1.7306713 6.6648828</span>
<span class="co">## year                                  34  1  0.5744677 0.5744677 1.4815830</span>
<span class="co">## month                                 34  4  1.3733243 0.3433311 0.8576774</span>
<span class="co">## day                                   34  3  1.1036246 0.3678749 0.9290946</span>
<span class="co">## subject                               34  1  0.7852501 0.7852501 2.0602011</span>
<span class="co">## reported.antibiotic.usage             34  1  0.5744677 0.5744677 1.4815830</span>
<span class="co">## days.since.experiment.start           34  4  1.3733243 0.3433311 0.8576774</span>
<span class="co">##                                     R2 Pr(&gt;F) AdjustedPvalue</span>
<span class="co">## sample.id                   1.00000000  1.000      1.0000000</span>
<span class="co">## barcode.sequence            1.00000000  1.000      1.0000000</span>
<span class="co">## body.site                   0.39993577  0.001      0.0090000</span>
<span class="co">## year                        0.04425069  0.126      0.2835000</span>
<span class="co">## month                       0.10578584  0.723      0.9295714</span>
<span class="co">## day                         0.08501112  0.532      0.9295714</span>
<span class="co">## subject                     0.06048705  0.052      0.2340000</span>
<span class="co">## reported.antibiotic.usage   0.04425069  0.116      0.2835000</span>
<span class="co">## days.since.experiment.start 0.10578584  0.718      0.9295714</span></code></pre>
</div>
<div class="section level3">
<h3 id="mantel-test-mantel">Mantel Test (MANTEL)<a class="anchor" aria-label="anchor" href="#mantel-test-mantel"></a>
</h3>
<p><span class="citation">(Mantel 1967)</span></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/run_MANTEL.html">run_MANTEL</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">ps_object</span>,
           y_variables <span class="op">=</span> <span class="st">"body.site"</span>,
           z_variables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"subject"</span>, <span class="st">"reported.antibiotic.usage"</span><span class="op">)</span>,
           norm <span class="op">=</span> <span class="cn">FALSE</span>,
           method <span class="op">=</span> <span class="st">"mantel.partial"</span>,
           method_dist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"bray"</span>, <span class="st">"euclidean"</span>, <span class="st">"jaccard"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">## Partial Mantel statistic based on Spearman's rank correlation rho </span>
<span class="co">## </span>
<span class="co">## Call:</span>
<span class="co">## vegan::mantel.partial(xdis = x_dis, ydis = y_dis, zdis = z_dis,      method = method_cor, permutations = 999) </span>
<span class="co">## </span>
<span class="co">## Mantel statistic r: 0.563 </span>
<span class="co">##       Significance: 0.001 </span>
<span class="co">## </span>
<span class="co">## Upper quantiles of permutations (null model):</span>
<span class="co">##    90%    95%  97.5%    99% </span>
<span class="co">## 0.0578 0.0779 0.0994 0.1285 </span>
<span class="co">## Permutation: free</span>
<span class="co">## Number of permutations: 999</span></code></pre>
</div>
<div class="section level3">
<h3 id="analysis-of-similarity-anosim">Analysis of Similarity (ANOSIM)<a class="anchor" aria-label="anchor" href="#analysis-of-similarity-anosim"></a>
</h3>
<p><span class="citation">(Clarke 1993)</span></p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/run_ANOSIM.html">run_ANOSIM</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">ps_object</span>,
           level <span class="op">=</span> <span class="st">"Genus"</span>,
           variable <span class="op">=</span> <span class="st">"body.site"</span>, 
           method <span class="op">=</span> <span class="st">"bray"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">## Call:</span>
<span class="co">## vegan::anosim(x = dis, grouping = datphe, permutations = 999) </span>
<span class="co">## Dissimilarity: bray </span>
<span class="co">## </span>
<span class="co">## ANOSIM statistic R: 0.6855 </span>
<span class="co">##       Significance: 0.001 </span>
<span class="co">## </span>
<span class="co">## Permutation: free</span>
<span class="co">## Number of permutations: 999</span></code></pre>
</div>
<div class="section level3">
<h3 id="multi-response-permutation-procedures-mrpp">Multi-response Permutation Procedures (MRPP)<a class="anchor" aria-label="anchor" href="#multi-response-permutation-procedures-mrpp"></a>
</h3>
<p><span class="citation">(Mielke Jr 1991)</span></p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/run_MRPP.html">run_MRPP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">ps_object</span>,
         level <span class="op">=</span> <span class="st">"Genus"</span>,
         variable <span class="op">=</span> <span class="st">"body.site"</span>, 
         method <span class="op">=</span> <span class="st">"bray"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">## Call:</span>
<span class="co">## vegan::mrpp(dat = dis, grouping = datphe, permutations = 999) </span>
<span class="co">## </span>
<span class="co">## Dissimilarity index: bray </span>
<span class="co">## Weights for groups:  n </span>
<span class="co">## </span>
<span class="co">## Class means and counts:</span>
<span class="co">## </span>
<span class="co">##       gut    left palm right palm tongue</span>
<span class="co">## delta 0.5657 0.7732    0.8839     0.5273</span>
<span class="co">## n     8      8         9          9     </span>
<span class="co">## </span>
<span class="co">## Chance corrected within-group agreement A: 0.2085 </span>
<span class="co">## Based on observed delta 0.6886 and expected delta 0.8699 </span>
<span class="co">## </span>
<span class="co">## Significance of delta: 0.001 </span>
<span class="co">## Permutation: free</span>
<span class="co">## Number of permutations: 999</span></code></pre>
</div>
<div class="section level3">
<h3 id="alpha-diversity">Alpha diversity<a class="anchor" aria-label="anchor" href="#alpha-diversity"></a>
</h3>
<ul>
<li>calculating alpha diversity index on microbiota data</li>
</ul>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">alphaindex</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_alphaindex.html">get_alphaindex</a></span><span class="op">(</span>
  ps <span class="op">=</span> <span class="va">ps_object</span>,
  level <span class="op">=</span> <span class="st">"Genus"</span>,
  indices <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Shannon"</span>, <span class="st">"Chao1"</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">alphaindex</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##        sample.id barcode.sequence body.site year month day   subject</span>
<span class="co">## L1S105    L1S105     AGTGCGATGCGT       gut 2009     3  17 subject-1</span>
<span class="co">## L1S140    L1S140     ATGGCAGCTCTA       gut 2008    10  28 subject-2</span>
<span class="co">## L1S208    L1S208     CTGAGATACGCG       gut 2009     1  20 subject-2</span>
<span class="co">## L1S257    L1S257     CCGACTGAGATG       gut 2009     3  17 subject-2</span>
<span class="co">## L1S281    L1S281     CCTCTCGTGATC       gut 2009     4  14 subject-2</span>
<span class="co">## L1S57      L1S57     ACACACTATGGC       gut 2009     1  20 subject-1</span>
<span class="co">##        reported.antibiotic.usage days.since.experiment.start Chao1  Shannon</span>
<span class="co">## L1S105                        No                         140    29 1.715803</span>
<span class="co">## L1S140                       Yes                           0    30 1.582306</span>
<span class="co">## L1S208                        No                          84    38 1.899839</span>
<span class="co">## L1S257                        No                         140    37 2.098328</span>
<span class="co">## L1S281                        No                         168    32 2.031926</span>
<span class="co">## L1S57                         No                          84    28 1.910027</span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="differential-analysis">Differential analysis<a class="anchor" aria-label="anchor" href="#differential-analysis"></a>
</h2>
<div class="section level3">
<h3 id="aldex">Aldex<a class="anchor" aria-label="anchor" href="#aldex"></a>
</h3>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ps_genus_group</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples</a></span><span class="op">(</span>
     <span class="va">ps_genus</span>,
     <span class="va">body.site</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gut"</span>, <span class="st">"tongue"</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="../reference/run_aldex.html">run_aldex</a></span><span class="op">(</span>ps <span class="op">=</span> <span class="va">ps_genus_group</span>, 
          group <span class="op">=</span> <span class="st">"body.site"</span>, 
          taxa_rank <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## microbiomeMarker-class inherited from phyloseq-class</span>
<span class="co">## normalization method:              [ none ]</span>
<span class="co">## microbiome marker identity method: [ ALDEx2_t.test ]</span>
<span class="co">## marker_table() Marker Table:       [ 36 microbiome markers with 5 variables ]</span>
<span class="co">## otu_table()    OTU Table:          [ 86 taxa and  17 samples ]</span>
<span class="co">## sample_data()  Sample Data:        [ 17 samples by  9 sample variables ]</span>
<span class="co">## tax_table()    Taxonomy Table:     [ 86 taxa by 1 taxonomic ranks ]</span></code></pre>
</div>
<div class="section level3">
<h3 id="limma-voom">limma-voom<a class="anchor" aria-label="anchor" href="#limma-voom"></a>
</h3>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/run_limma_voom.html">run_limma_voom</a></span><span class="op">(</span>ps <span class="op">=</span> <span class="va">ps_genus_group</span>, 
               group <span class="op">=</span> <span class="st">"body.site"</span>, 
               taxa_rank <span class="op">=</span> <span class="st">"Phylum"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## microbiomeMarker-class inherited from phyloseq-class</span>
<span class="co">## normalization method:              [ none ]</span>
<span class="co">## microbiome marker identity method: [ limma_voom ]</span>
<span class="co">## marker_table() Marker Table:       [ 8 microbiome markers with 5 variables ]</span>
<span class="co">## otu_table()    OTU Table:          [ 8 taxa and  17 samples ]</span>
<span class="co">## sample_data()  Sample Data:        [ 17 samples by  9 sample variables ]</span>
<span class="co">## tax_table()    Taxonomy Table:     [ 8 taxa by 1 taxonomic ranks ]</span></code></pre>
</div>
<div class="section level3">
<h3 id="ancom">ANCOM<a class="anchor" aria-label="anchor" href="#ancom"></a>
</h3>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"enterotypes_arumugam"</span><span class="op">)</span>
<span class="va">ps_enterotypes</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples</a></span><span class="op">(</span>
     <span class="va">enterotypes_arumugam</span>,
     <span class="va">Enterotype</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Enterotype 3"</span>, <span class="st">"Enterotype 2"</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="../reference/run_ancom.html">run_ancom</a></span><span class="op">(</span>ps <span class="op">=</span> <span class="va">ps_enterotypes</span>, 
          group <span class="op">=</span> <span class="st">"Enterotype"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## microbiomeMarker-class inherited from phyloseq-class</span>
<span class="co">## normalization method:              [ TSS ]</span>
<span class="co">## microbiome marker identity method: [ ANCOM ]</span>
<span class="co">## marker_table() Marker Table:       [ 13 microbiome markers with 4 variables ]</span>
<span class="co">## otu_table()    OTU Table:          [ 235 taxa and  24 samples ]</span>
<span class="co">## sample_data()  Sample Data:        [ 24 samples by  9 sample variables ]</span>
<span class="co">## tax_table()    Taxonomy Table:     [ 235 taxa by 1 taxonomic ranks ]</span></code></pre>
</div>
<div class="section level3">
<h3 id="ancombc">ANCOMBC<a class="anchor" aria-label="anchor" href="#ancombc"></a>
</h3>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/run_ancombc.html">run_ancombc</a></span><span class="op">(</span>ps <span class="op">=</span> <span class="va">ps_enterotypes</span>, 
            group <span class="op">=</span> <span class="st">"Enterotype"</span>, 
            confounders <span class="op">=</span> <span class="st">"Gender"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## microbiomeMarker-class inherited from phyloseq-class</span>
<span class="co">## normalization method:              [ none ]</span>
<span class="co">## microbiome marker identity method: [ ancombc ]</span>
<span class="co">## marker_table() Marker Table:       [ no microbiome markers were identified ]</span>
<span class="co">## otu_table()    OTU Table:          [ 235 taxa and  24 samples ]</span>
<span class="co">## sample_data()  Sample Data:        [ 24 samples by  9 sample variables ]</span>
<span class="co">## tax_table()    Taxonomy Table:     [ 235 taxa by 1 taxonomic ranks ]</span></code></pre>
</div>
<div class="section level3">
<h3 id="deseq2">DESeq2<a class="anchor" aria-label="anchor" href="#deseq2"></a>
</h3>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/run_deseq2.html">run_deseq2</a></span><span class="op">(</span>ps <span class="op">=</span> <span class="va">ps_enterotypes</span>, 
           group <span class="op">=</span> <span class="st">"Enterotype"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## microbiomeMarker-class inherited from phyloseq-class</span>
<span class="co">## normalization method:              [ RLE ]</span>
<span class="co">## microbiome marker identity method: [ DESeq2: Wald ]</span>
<span class="co">## marker_table() Marker Table:       [ 16 microbiome markers with 5 variables ]</span>
<span class="co">## otu_table()    OTU Table:          [ 235 taxa and  24 samples ]</span>
<span class="co">## sample_data()  Sample Data:        [ 24 samples by  10 sample variables ]</span>
<span class="co">## tax_table()    Taxonomy Table:     [ 235 taxa by 1 taxonomic ranks ]</span></code></pre>
</div>
<div class="section level3">
<h3 id="edger">edgeR<a class="anchor" aria-label="anchor" href="#edger"></a>
</h3>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/run_edger.html">run_edger</a></span><span class="op">(</span><span class="va">ps_enterotypes</span>, group <span class="op">=</span> <span class="st">"Enterotype"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## microbiomeMarker-class inherited from phyloseq-class</span>
<span class="co">## normalization method:              [ TMM ]</span>
<span class="co">## microbiome marker identity method: [ edgeR: LRT ]</span>
<span class="co">## marker_table() Marker Table:       [ 34 microbiome markers with 5 variables ]</span>
<span class="co">## otu_table()    OTU Table:          [ 235 taxa and  24 samples ]</span>
<span class="co">## sample_data()  Sample Data:        [ 24 samples by  10 sample variables ]</span>
<span class="co">## tax_table()    Taxonomy Table:     [ 235 taxa by 1 taxonomic ranks ]</span></code></pre>
</div>
<div class="section level3">
<h3 id="lefse">lefse<a class="anchor" aria-label="anchor" href="#lefse"></a>
</h3>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/run_lefse.html">run_lefse</a></span><span class="op">(</span>ps <span class="op">=</span> <span class="va">ps_enterotypes</span>, 
          group <span class="op">=</span> <span class="st">"Enterotype"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## microbiomeMarker-class inherited from phyloseq-class</span>
<span class="co">## normalization method:              [ CPM ]</span>
<span class="co">## microbiome marker identity method: [ lefse ]</span>
<span class="co">## marker_table() Marker Table:       [ 14 microbiome markers with 5 variables ]</span>
<span class="co">## otu_table()    OTU Table:          [ 235 taxa and  24 samples ]</span>
<span class="co">## sample_data()  Sample Data:        [ 24 samples by  9 sample variables ]</span>
<span class="co">## tax_table()    Taxonomy Table:     [ 235 taxa by 1 taxonomic ranks ]</span></code></pre>
</div>
<div class="section level3">
<h3 id="metagenomeseq">metagenomeSeq<a class="anchor" aria-label="anchor" href="#metagenomeseq"></a>
</h3>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/run_metagenomeseq.html">run_metagenomeseq</a></span><span class="op">(</span>ps <span class="op">=</span> <span class="va">ps_enterotypes</span>, 
                  group <span class="op">=</span> <span class="st">"Enterotype"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## microbiomeMarker-class inherited from phyloseq-class</span>
<span class="co">## normalization method:              [ CSS ]</span>
<span class="co">## microbiome marker identity method: [ metagenomeSeq: ZILN ]</span>
<span class="co">## marker_table() Marker Table:       [ 16 microbiome markers with 5 variables ]</span>
<span class="co">## otu_table()    OTU Table:          [ 235 taxa and  24 samples ]</span>
<span class="co">## sample_data()  Sample Data:        [ 24 samples by  10 sample variables ]</span>
<span class="co">## tax_table()    Taxonomy Table:     [ 235 taxa by 1 taxonomic ranks ]</span></code></pre>
</div>
<div class="section level3">
<h3 id="test_multiple_groups">test_multiple_groups<a class="anchor" aria-label="anchor" href="#test_multiple_groups"></a>
</h3>
<ul>
<li><p>anova</p></li>
<li><p>kruskal</p></li>
</ul>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ps</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples</a></span><span class="op">(</span>
    <span class="va">enterotypes_arumugam</span>,
    <span class="va">Enterotype</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Enterotype 3"</span>, <span class="st">"Enterotype 2"</span>, <span class="st">"Enterotype 1"</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="../reference/run_test_multiple_groups.html">run_test_multiple_groups</a></span><span class="op">(</span>ps <span class="op">=</span> <span class="va">ps</span>, 
                         group <span class="op">=</span> <span class="st">"Enterotype"</span>, 
                         method <span class="op">=</span> <span class="st">"anova"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## microbiomeMarker-class inherited from phyloseq-class</span>
<span class="co">## normalization method:              [ TSS ]</span>
<span class="co">## microbiome marker identity method: [ anova ]</span>
<span class="co">## marker_table() Marker Table:       [ 24 microbiome markers with 5 variables ]</span>
<span class="co">## otu_table()    OTU Table:          [ 238 taxa and  32 samples ]</span>
<span class="co">## sample_data()  Sample Data:        [ 32 samples by  9 sample variables ]</span>
<span class="co">## tax_table()    Taxonomy Table:     [ 238 taxa by 1 taxonomic ranks ]</span></code></pre>
</div>
<div class="section level3">
<h3 id="supervised-leaning-sl-methods">supervised leaning (SL) methods<a class="anchor" aria-label="anchor" href="#supervised-leaning-sl-methods"></a>
</h3>
<ul>
<li>logistic regression</li>
</ul>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ps_small</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_taxa-methods.html" class="external-link">subset_taxa</a></span><span class="op">(</span>
    <span class="va">enterotypes_arumugam</span>,
    <span class="va">Phylum</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Firmicutes"</span>, <span class="st">"Bacteroidetes"</span><span class="op">)</span>
<span class="op">)</span>

<span class="va">mm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_sl.html">run_sl</a></span><span class="op">(</span>
    ps <span class="op">=</span> <span class="va">ps_small</span>,
    group <span class="op">=</span> <span class="st">"Gender"</span>,
    taxa_rank <span class="op">=</span> <span class="st">"Genus"</span>,
    nfolds <span class="op">=</span> <span class="fl">2</span>,
    nrepeats <span class="op">=</span> <span class="fl">1</span>,
    top_n <span class="op">=</span> <span class="fl">15</span>,
    norm <span class="op">=</span> <span class="st">"TSS"</span>,
    method <span class="op">=</span> <span class="st">"LR"</span><span class="op">)</span>

<span class="va">mm</span></code></pre></div>
<pre><code><span class="co">## microbiomeMarker-class inherited from phyloseq-class</span>
<span class="co">## normalization method:              [ TSS ]</span>
<span class="co">## microbiome marker identity method: [ logistic regression ]</span>
<span class="co">## marker_table() Marker Table:       [ 15 microbiome markers with 3 variables ]</span>
<span class="co">## otu_table()    OTU Table:          [ 91 taxa and  39 samples ]</span>
<span class="co">## sample_data()  Sample Data:        [ 39 samples by  9 sample variables ]</span>
<span class="co">## tax_table()    Taxonomy Table:     [ 91 taxa by 1 taxonomic ranks ]</span></code></pre>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<p>This vignette was created under the following conditions:</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R version 4.1.2 (2021-11-01)</span>
<span class="co">## Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">## Running under: macOS Big Sur 10.16</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib</span>
<span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">## [1] caret_6.0-92             lattice_0.20-45          conflicted_1.1.0        </span>
<span class="co">## [4] MicrobiomeAnalysis_1.0.3 tibble_3.1.8             dplyr_1.0.10            </span>
<span class="co">## [7] phyloseq_1.38.0          ggplot2_3.4.0            knitr_1.40              </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] systemfonts_1.0.4           plyr_1.8.6                 </span>
<span class="co">##   [3] igraph_1.2.11               splines_4.1.2              </span>
<span class="co">##   [5] listenv_0.8.0               BiocParallel_1.28.3        </span>
<span class="co">##   [7] GenomeInfoDb_1.30.1         digest_0.6.30              </span>
<span class="co">##   [9] foreach_1.5.2               htmltools_0.5.3            </span>
<span class="co">##  [11] fansi_1.0.2                 magrittr_2.0.3             </span>
<span class="co">##  [13] memoise_2.0.1               cluster_2.1.2              </span>
<span class="co">##  [15] limma_3.50.1                globals_0.16.1             </span>
<span class="co">##  [17] recipes_1.0.1               Biostrings_2.62.0          </span>
<span class="co">##  [19] annotate_1.72.0             gower_1.0.0                </span>
<span class="co">##  [21] matrixStats_0.61.0          hardhat_1.2.0              </span>
<span class="co">##  [23] metagenomeSeq_1.36.0        pkgdown_2.0.3              </span>
<span class="co">##  [25] colorspace_2.0-3            blob_1.2.2                 </span>
<span class="co">##  [27] textshaping_0.3.6           rbibutils_2.2.7            </span>
<span class="co">##  [29] xfun_0.34                   crayon_1.5.0               </span>
<span class="co">##  [31] RCurl_1.98-1.6              jsonlite_1.8.3             </span>
<span class="co">##  [33] genefilter_1.76.0           survival_3.4-0             </span>
<span class="co">##  [35] iterators_1.0.14            ape_5.6-2                  </span>
<span class="co">##  [37] glue_1.6.2                  gtable_0.3.0               </span>
<span class="co">##  [39] ipred_0.9-12                zlibbioc_1.40.0            </span>
<span class="co">##  [41] XVector_0.34.0              DelayedArray_0.20.0        </span>
<span class="co">##  [43] RcppZiggurat_0.1.6          Rhdf5lib_1.16.0            </span>
<span class="co">##  [45] future.apply_1.8.1          shape_1.4.6                </span>
<span class="co">##  [47] BiocGenerics_0.40.0         scales_1.2.1               </span>
<span class="co">##  [49] edgeR_3.36.0                DBI_1.1.2                  </span>
<span class="co">##  [51] Rcpp_1.0.9                  xtable_1.8-4               </span>
<span class="co">##  [53] bit_4.0.4                   lava_1.6.10                </span>
<span class="co">##  [55] prodlim_2019.11.13          stats4_4.1.2               </span>
<span class="co">##  [57] truncnorm_1.0-8             glmnet_4.1-3               </span>
<span class="co">##  [59] httr_1.4.2                  gplots_3.1.1               </span>
<span class="co">##  [61] RColorBrewer_1.1-2          ellipsis_0.3.2             </span>
<span class="co">##  [63] NADA_1.6-1.1                pkgconfig_2.0.3            </span>
<span class="co">##  [65] XML_3.99-0.9                nnet_7.3-17                </span>
<span class="co">##  [67] sass_0.4.2                  locfit_1.5-9.5             </span>
<span class="co">##  [69] utf8_1.2.2                  tidyselect_1.1.2           </span>
<span class="co">##  [71] rlang_1.0.6                 reshape2_1.4.4             </span>
<span class="co">##  [73] AnnotationDbi_1.56.2        munsell_0.5.0              </span>
<span class="co">##  [75] tools_4.1.2                 cachem_1.0.6               </span>
<span class="co">##  [77] cli_3.4.1                   generics_0.1.2             </span>
<span class="co">##  [79] RSQLite_2.2.10              ade4_1.7-18                </span>
<span class="co">##  [81] evaluate_0.17               biomformat_1.22.0          </span>
<span class="co">##  [83] stringr_1.4.1               fastmap_1.1.0              </span>
<span class="co">##  [85] yaml_2.3.6                  ragg_1.2.2                 </span>
<span class="co">##  [87] ModelMetrics_1.2.2.2        bit64_4.0.5                </span>
<span class="co">##  [89] fs_1.5.2                    caTools_1.18.2             </span>
<span class="co">##  [91] purrr_0.3.4                 KEGGREST_1.34.0            </span>
<span class="co">##  [93] future_1.28.0               ANCOMBC_1.4.0              </span>
<span class="co">##  [95] nlme_3.1-155                compiler_4.1.2             </span>
<span class="co">##  [97] rstudioapi_0.13             png_0.1-7                  </span>
<span class="co">##  [99] zCompositions_1.4.0         geneplotter_1.72.0         </span>
<span class="co">## [101] bslib_0.4.0                 stringi_1.7.8              </span>
<span class="co">## [103] highr_0.9                   desc_1.4.1                 </span>
<span class="co">## [105] Matrix_1.4-0                nloptr_2.0.0               </span>
<span class="co">## [107] vegan_2.5-7                 microbiome_1.16.0          </span>
<span class="co">## [109] permute_0.9-7               multtest_2.50.0            </span>
<span class="co">## [111] vctrs_0.5.1                 pillar_1.7.0               </span>
<span class="co">## [113] lifecycle_1.0.3             rhdf5filters_1.6.0         </span>
<span class="co">## [115] Rdpack_2.2                  jquerylib_0.1.4            </span>
<span class="co">## [117] data.table_1.14.6           bitops_1.0-7               </span>
<span class="co">## [119] GenomicRanges_1.46.1        R6_2.5.1                   </span>
<span class="co">## [121] KernSmooth_2.23-20          parallelly_1.32.1          </span>
<span class="co">## [123] IRanges_2.28.0              codetools_0.2-18           </span>
<span class="co">## [125] MASS_7.3-55                 gtools_3.9.2               </span>
<span class="co">## [127] assertthat_0.2.1            Wrench_1.12.0              </span>
<span class="co">## [129] rhdf5_2.38.1                SummarizedExperiment_1.24.0</span>
<span class="co">## [131] DESeq2_1.34.0               rprojroot_2.0.2            </span>
<span class="co">## [133] withr_2.5.0                 S4Vectors_0.32.3           </span>
<span class="co">## [135] GenomeInfoDbData_1.2.7      mgcv_1.8-39                </span>
<span class="co">## [137] parallel_4.1.2              rpart_4.1.16               </span>
<span class="co">## [139] timeDate_3043.102           grid_4.1.2                 </span>
<span class="co">## [141] prettydoc_0.4.1             class_7.3-20               </span>
<span class="co">## [143] tidyr_1.2.0                 Rfast_2.0.6                </span>
<span class="co">## [145] rmarkdown_2.17              MatrixGenerics_1.6.0       </span>
<span class="co">## [147] Rtsne_0.15                  pROC_1.18.0                </span>
<span class="co">## [149] lubridate_1.8.0             ALDEx2_1.30.0              </span>
<span class="co">## [151] Biobase_2.54.0</span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-anderson2014permutational" class="csl-entry">
Anderson, Marti J. 2014. <span>“Permutational Multivariate Analysis of
Variance (PERMANOVA).”</span> <em>Wiley Statsref: Statistics Reference
Online</em>, 1–15.
</div>
<div id="ref-anderson2006multivariate" class="csl-entry">
Anderson, Marti J, Kari E Ellingsen, and Brian H McArdle. 2006.
<span>“Multivariate Dispersion as a Measure of Beta Diversity.”</span>
<em>Ecology Letters</em> 9 (6): 683–93.
</div>
<div id="ref-clarke1993non" class="csl-entry">
Clarke, K Robert. 1993. <span>“Non-Parametric Multivariate Analyses of
Changes in Community Structure.”</span> <em>Australian Journal of
Ecology</em> 18 (1): 117–43.
</div>
<div id="ref-mantel1967detection" class="csl-entry">
Mantel, Nathan. 1967. <span>“The Detection of Disease Clustering and a
Generalized Regression Approach.”</span> <em>Cancer Research</em> 27 (2
Part 1): 209–20.
</div>
<div id="ref-mielke1991application" class="csl-entry">
Mielke Jr, Paul W. 1991. <span>“The Application of Multivariate
Permutation Methods Based on Distance Functions in the Earth
Sciences.”</span> <em>Earth-Science Reviews</em> 31 (1): 55–71.
</div>
<div id="ref-thingholm2019obese" class="csl-entry">
Thingholm, Louise B, Malte C Rühlemann, Manja Koch, Brie Fuqua, Guido
Laucke, Ruwen Boehm, Corinna Bang, et al. 2019. <span>“Obese Individuals
with and Without Type 2 Diabetes Show Different Gut Microbial Functional
Capacity and Composition.”</span> <em>Cell Host &amp; Microbe</em> 26
(2): 252–64.
</div>
<div id="ref-xia2018statistical" class="csl-entry">
Xia, Yinglin, Jun Sun, Ding-Geng Chen, et al. 2018. <em>Statistical
Analysis of Microbiome Data with r</em>. Vol. 847. Springer.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Hua Zou.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
